File: tools/news/lib/news.js
'use strict';

const disallowDomains = [
  'medium.com',
];

class NewsPosts {
  apiURI = 'https://hacker-news.firebaseio.com/v0/';
  loadLimit = 30;
  allPosts = null;
  genPost = null;
  loadedPosts = [];

  async init() {
    this.allPosts = JSON.parse((await this.getTop500()));
    this.genPost = NewsPosts.postList(this.allPosts);
  }

  async getTop500() {
    const res = await fetch(`${this.apiURI}topstories.json`);
    return await res.text();
  }

  async getPostData(id) {
    const res = await fetch(`${this.apiURI}item/${id}.json`);
    return await res.text();
  }

  recordPosted(story) {
    this.loadedPosts.push(story);
  }

  static formatLayout(title, url, id) {
    const urlMatcher = /:\/\/(?:www\.)?(.[^/]+)/;
    const pocketUrl = `https://getpocket.com/save?url=${url}`;
    const commentsUrl = `https://news.ycombinator.com/item?id=${id}`
    return `<hr><p>
      <a href="${url}">${title}</a>
      <br />
      <span class="source">${url ? url.match(urlMatcher)[1] : ''}</span>
      &nbsp;<a class="extra-content-link" href="${pocketUrl}" target="_blank" rel="noopener noreferrer">add to pocket</a>
      &nbsp;<a class="extra-content-link" href="${commentsUrl} target="_blank" rel="noopener noreferrer">comments</a>
    </p>`;
  }

  static *postList(lists) {
    while (lists.length) yield lists.shift();
  }
}

const run = async (posts) => {
  try {
    await posts.init();
    await loadSomePosts(posts);
  } catch (e) {
    console.error(e);
    alert('An error occured. Please try again later');
  }
};

const loadSomePosts = async (posts) => {
  showSpinner();
  try {
    const list = await loadPosts(posts.loadLimit, posts);
    for (let story of list) {
      story = JSON.parse(story) ?? {};
      const shortUrlMatch = (story.url ?? '').match(/:\/\/(?:www\.)?(.[^/]+)/);
      if (story.url && shortUrlMatch.length & 2
          && !disallowDomains.includes(shortUrlMatch[1])) {
        injectPost(story.title, story.url, story.id);
        posts.recordPosted(story);
      } else {
        continue;
      }
    }
  } finally {
    hideSpinner();
    showFooter();
  }
};

const hideSpinner = () => {
  document.getElementById('spinner').style.display = 'none';
};

const showSpinner = () => {
  document.getElementById('spinner').style.display = 'block';
};

const showFooter = () => {
  document.getElementById('footer').style.display = 'block';
};

const loadPosts = async (numPostsToLoad, posts) => {
  let firstPostsFuture = [];

  for (let i = 0; i < numPostsToLoad; i++) {
    firstPostsFuture.push(posts.getPostData(posts.genPost.next().value));
  }
  const firstPostsList = await Promise.all(firstPostsFuture);
  return firstPostsList;
};

const injectPost = (title, url, id) => {
  const layout = NewsPosts.formatLayout(title, url, id);
  const div = document.createElement('div');
  div.className = 'post';
  div.innerHTML = layout;
  const main = document.querySelector('main.content');
  if (main) {
    main.append(div);
  }
};

const addUpdatedMessage = () => {
  const msg = document.querySelector('h6.updated-message');
  window.newsUpdateTime = Date.now();
  msg.innerText = 'Last updated now';

  window.newsUpdateTimer = setInterval(() => {
    const timeNow = Date.now();
    const timeSince = (timeNow - window.newsUpdateTime) / 60000;
    if (timeSince < 60) {
      const unit = timeSince === 1 ? 'minute' : 'minutes';
      msg.innerText = `Last updated ${timeSince | 0} ${unit} ago`;
    } else {
      clearInterval(window.newsUpdateTimer);
      msg.innerText = `Last updated 1 hour ago`;
      window.newsUpdateTimer = setInterval(() => {
        const timeNow = Date.now();
        const timeSince = (timeNow - window.newsUpdateTime) / 3600000;
        const unit = timeSince === 1 ? 'hour' : 'hours';
        msg.innerText = `Last updated ${timeSince | 0} ${unit} ago`;
      });
    }
  }, 300000);
};

const saveToPocket = (url) => {
  const pocketUrl = `https://getpocket.com/save?url=${url}`;
  window.open(pocketUrl, '_blank');
};

const main = () => {
	const posts = new NewsPosts();
	document.addEventListener('DOMContentLoaded', run.bind(null, posts));
	document.addEventListener('scroll', evt => {
		if (window.innerHeight + window.scrollY >= document.body.offsetHeight) {
			loadSomePosts(posts);
		}
	});
	addUpdatedMessage();
  document.querySelector('button.open-all').addEventListener('click', () => {
    posts.loadedPosts.forEach((p) => window.open(p.url, '_blank'));
  });
};

// run script
main();

File: jmthornton/news.py
"""News functions."""

import requests


API_URL = "https://hacker-news.firebaseio.com/v0"
TOP_POSTS = f"{API_URL}/topstories.json"
POST_DETAILS = f"{API_URL}/item/{{}}.json"
DISALLOWED_DOMAINS = ["medium.com"]
LOAD_LIMIT = 50
POCKET_URL = "https://getpocket.com/save?url={}"
COMMENTS_URL = "https://news.ycombinator.com/item?id={}"


def get_news_posts():
    """Get news posts."""
    all_posts_ids = fetch_top_post_ids()
    if not all_posts_ids:
        return []
    posts = get_all_posts_details(all_posts_ids)
    return posts


def fetch_top_post_ids():
    """Get top stories from API."""
    # TODO cache
    top_posts = requests.get(TOP_POSTS)
    if top_posts.status_code != 200:
        return None
    return top_posts.json()


def get_all_posts_details(post_ids):
    """Get posts details."""
    posts = []
    # TODO async
    for post_id in post_ids:
        res = requests.get(POST_DETAILS.format(post_id))
        if res.status_code != 200:
            continue
        post = res.json()
        if is_acceptable(post):
            post = add_metadata(post)
            posts.append(post)
        if len(posts) >= LOAD_LIMIT:
            break

    return posts


def is_acceptable(post):
    """Check if post is acceptable."""
    if not post.get("url") or any(
        domain in post["url"] for domain in DISALLOWED_DOMAINS
    ):
        return False
    return "title" in post and "id" in post


def add_metadata(post):
    """Add metadata."""
    post["source"] = post["url"].split("/")[2]
    post["pocket_url"] = POCKET_URL.format(post["url"])
    post["comments_url"] = COMMENTS_URL.format(post["id"])
    return post
